lib/Lower/Mangler.cpp

```diff
@@ -19,32 +19,37 @@
 namespace {
 
 // recursively build the vector of module scopes
-void moduleNames(const Fortran::semantics::Scope *scope,
+void moduleNames(const Fortran::semantics::Scope &scope,
                  llvm::SmallVector<llvm::StringRef, 2> &result) {
-  if (scope->kind() == Fortran::semantics::Scope::Kind::Global) {
+  if (scope.kind() == Fortran::semantics::Scope::Kind::Global) {
     return;
   }
-  moduleNames(&scope->parent(), result);
-  if (scope->kind() == Fortran::semantics::Scope::Kind::Module)
-    if (auto *symbol = scope->symbol())
+  moduleNames(scope.parent(), result);
+  if (scope.kind() == Fortran::semantics::Scope::Kind::Module)
+    if (auto *symbol = scope.symbol())
       result.emplace_back(toStringRef(symbol->name()));
 }
 
 llvm::SmallVector<llvm::StringRef, 2>
-moduleNames(const Fortran::semantics::Scope *scope) {
+moduleNames(const Fortran::semantics::Scope &scope) {
   llvm::SmallVector<llvm::StringRef, 2> result;
   moduleNames(scope, result);
   return result;
 }
 
 llvm::Optional<llvm::StringRef>
-hostName(const Fortran::semantics::Scope *scope) {
-  if (scope->kind() == Fortran::semantics::Scope::Kind::Subprogram) {
-    auto &parent = scope->parent();
-    if (parent.kind() == Fortran::semantics::Scope::Kind::Subprogram)
-      if (auto *symbol = parent.symbol()) {
-        return {toStringRef(symbol->name())};
-      }
+hostName(const Fortran::semantics::Scope &scope) {
+  //  if (scope.kind() == Fortran::semantics::Scope::Kind::Subprogram) {
+  //    auto &parent = scope.parent();
+  //    if (parent.kind() == Fortran::semantics::Scope::Kind::Subprogram)
+  //      if (auto *symbol = parent.symbol()) {
+  //        return {toStringRef(symbol->name())};
+  //      }
+  //  }
+  //  return {};
+  if (scope.kind() == Fortran::semantics::Scope::Kind::Subprogram) {
+    assert(scope.symbol() && "subprogram scope must have a symbol");
+    return {toStringRef(scope.symbol()->name())};
```

Comment:
Something looks awry here. It looks like `hostName` returned the name of the host PROCEDURE. It looks like this change returns the name of the internal PROCEDURE.

Reply:
This change only makes sens in light of another subtle other change in `mangleName`, the scope is now coming from `symbol->owner()` instead of `symbol->scope()`. The difference is that the `symbol->scope` returned the scope of the function being mangled when `symbol->owner()` returns the scope that defines the function.

So in the case of an internal procedure, the scope we are providing to `hostName` is directly the host procedure scope, so there is no need to look one scope up here anymore.
I've asked Tim and I'll just fix related semantic to add the flag.

Reply:
Aside: we should remove the old code instead of commenting it out.

Reply:
Yup, I removed that, I usually keep it while refactoring to proof read the changes.